### PR TITLE
invoice hook

### PR DESCRIPTION
**Pull request by rustyrussell**
So many hooks to write, so little time...

### CODE REVIEW

**Comment by rustyrussell** [comment]
Fixed test I accidentally broke, python flake8 fails, and rebased...

File: lightningd/peer_htlcs.c [hunk 1]
```diff
@@ -216,12 +224,19 @@ static bool check_cltv(struct htlc_in *hin,
...
 }
 
-static void fulfill_htlc(struct htlc_in *hin, const struct preimage *preimage)
+void fulfill_htlc(struct htlc_in *hin, const struct preimage *preimage)
 {
 	u8 *msg;
 	struct channel *channel = hin->key.channel;
 	struct wallet *wallet = channel->peer->ld->wallet;
 
+	if (hin->hstate != RCVD_ADD_ACK_REVOCATION) {
```
Comment (cdecker) [inline lightningd/peer_htlcs.c]: I assume this may happen due to the updates no longer being atomic?

File: lightningd/invoice.c [hunk 2]
```diff
@@ -99,13 +100,125 @@ static void wait_on_invoice(const struct invoice *invoice, void *cmd)
...
+	t = json_get_member(buffer, resulttok, "failure_code");
+	if (!t)
+		return 0;
+
+	if (!json_to_number(buffer, t, &val))
+		fatal("Invalid invoice_payment_hook failure_code: %.*s",
+		      toks[0].end - toks[1].start, buffer);
+
+	/* UPDATE isn't valid for final nodes to return, and I think we
+	 * assert elsewhere that we don't do this! */
```
Comment (cdecker) [inline lightningd/invoice.c]: Yep, just stumbled over the check before :+1:

**Comment by cdecker** [comment]
ACK https://github.com/ElementsProject/lightning/pull/2540/commits/9a420fe09376875a4f28d90346bf7005d2840dc6

File: lightningd/peer_htlcs.c (hunk 1 above)
Comment (rustyrussell) [reply lightningd/peer_htlcs.c]: Yes, in particular we might have gone onchain while waiting, then timed out the HTLC.

**Pull request merged**